src/components/input/input.tsx

```diff
@@ -955,19 +953,18 @@ export class Input
       const newLocalizedValue = numberStringFormatter.localize(newValue);
       this.localizedValue = newLocalizedValue;
 
-      this.setPreviousValue(previousValue || this.value);
-      this.previousValueOrigin = origin;
       this.userChangedValue = origin === "user" && this.value !== newValue;
       // don't sanitize the start of negative/decimal numbers, but
       // don't set value to an invalid number
       this.value = ["-", "."].includes(newValue) ? "" : newValue;
-      origin === "direct" && this.setInputValue(newLocalizedValue);
     } else {
-      this.setPreviousValue(previousValue || this.value);
-      this.previousValueOrigin = origin;
       this.userChangedValue = origin === "user" && this.value !== value;
       this.value = value;
-      origin === "direct" && this.setInputValue(value);
+    }
+
+    if (origin === "direct") {
+      this.setInputValue(value);
+      this.previousEmittedValue = value;
```

Comment:
Setting previousEmittedValue here is also crucial to do in the direct case.  Glad to see the `origin` pattern being put to good use!